cql3/expr/prepare_expr.cc

```diff
@@ -817,17 +817,36 @@ cast_prepare_expression(const cast& c, data_dictionary::database db, const sstri
 
 std::optional<expression>
 prepare_function_call(const expr::function_call& fc, data_dictionary::database db, const sstring& keyspace, const schema* schema_opt, lw_shared_ptr<column_specification> receiver) {
-    if (!receiver) {
-        // TODO: It is possible to infer the type of a function call if there is only one overload, or if all overloads return the same type
+    if (!schema_opt && !receiver) {
+        // In order to prepare a function_call we need to know the keyspace and column family (table name),
+        // which define the environment where this function is defined and executed.
+        // The information about keyspace is provided as an argument, but the information about column_family
+        // has to be acquired either from schema_opt or receiver.
+        // When both of them are not available it's impossible to prepare the function.
         return std::nullopt;
     }
+    const sstring& cf_name = schema_opt ? schema_opt->cf_name() : receiver->cf_name;
+
+    // Prepare the arguments that can be prepared without a receiver.
+    // Prepared expressions have a known type, which helps with finding the right function.
+    std::vector<expression> partially_prepared_args;
+    for (const expression& argument : fc.args) {
+        std::optional<expression> prepared_arg_opt = try_prepare_expression(argument, db, keyspace, schema_opt, nullptr);
+        if (prepared_arg_opt.has_value()) {
+            partially_prepared_args.emplace_back(*prepared_arg_opt);
+        } else {
+            partially_prepared_args.push_back(argument);
```

Comment:
I feel that it's wrong to lose information about whether an argument was prepared or not. Won't we prepare it again?

Reply:
I added this because I ran into problems with `unresolved_identifier` and `constant`. `functions::get` requires a list of function arguments, where each argument is `assignment_testable`. `unresolved_identifier` can't be `assignment_testable`, because it represents an unknown column, there's just the column name.

To fix it I prepare the arguments that can be prepared without a receiver, which includes `unresolved_identifiers`. I could potentially prepare only the occurrences of `unresolved_identifier`, but this felt like a more general solution.

We will prepare some of them again, but is it wrong to do so?
I took a look at `prepare_expression` and it looks like it's able to reprepare a previously prepared expression for all the cases that matter.

I think that `prepare_expression` should be robust enough that calling it twice isn't a bug.  It's hard to keep track of which expressions were already prepared and which ones were not, I bet we already call prepare twice somewhere by accident.